src/pykeen/training/training_loop.py

```diff
@@ -415,6 +417,90 @@ def train(
 
         return result
 
+    def _train_epoch(
+        self, 
+        batches: DataLoader, 
+        callbacks: MultiTrainingCallback, 
+        sub_batch_size: Optional[int],
+        label_smoothing: float,
+        slice_size: Optional[int],
+        epoch: int,
+        only_size_probing: bool,
+    ) -> float:
+        """
+        Run one epoch.
+        
+        :param batches:
+            the batches to process
+        :param callbacks:
+            the callbacks to apply (wrapped into a single object)
+        :param sub_batch_size:
+            the sub-batch size to use
+        :param label_smoothing:
+            the label-smoothing to apply
+        :param slice_size:
+            the optional slice size
+        :param epoch:
+            the current epoch (only used to forward to callbacks)
+        :param only_size_probing:
+            whether to stop after the second batch
+        
+        :return:
+            the epoch loss
+        """
+        # Accumulate loss over epoch
+        current_epoch_loss = 0.0
+
+        # Flag to check when to quit the size probing
+        evaluated_once = False
+
+        for batch in batches:
+            # apply callbacks before starting with batch
+            callbacks.pre_batch()
+
+            # Get batch size of current batch (last batch may be incomplete)
+            current_batch_size = self._get_batch_size(batch)
+            _sub_batch_size = sub_batch_size or current_batch_size
+
+            # accumulate gradients for whole batch
+            for start in range(0, current_batch_size, _sub_batch_size):
+                stop = min(start + _sub_batch_size, current_batch_size)
+
+                # forward pass call
+                batch_loss = self._forward_pass(
+                    batch,
+                    start,
+                    stop,
+                    current_batch_size,
+                    label_smoothing,
+                    slice_size,
+                )
+                current_epoch_loss += batch_loss
+                callbacks.on_batch(epoch=epoch, batch=batch, batch_loss=batch_loss)
+
+            callbacks.pre_step()
```

Comment:
is there a step? strange that these two go in a row with anything happening in the middle

Reply:
This is an artifact of gradient clippers being callbacks; they need to be applied *before* the optimizer callback's `post_batch`.

Either moving the clippers into the optimizer callback, or having a way to (robustly) define precedence between callbacks could help.

Reply:
hmm well I guess there's no harm in keeping these extra names at least for specificity